torch/lib/c10d/reducer.cpp

```diff
@@ -455,31 +520,38 @@ void Reducer::autograd_hook(VariableIndex index) {
     local_used_maps_[index.replica_index][index.variable_index] = 1;
   }
 
-  // Rebuild bucket only if 1) it is the first time to rebuild bucket 2)
-  // find_unused_parameters_ is false, currently it does not support when there
-  // are unused parameters 3) this backward pass needs to run allreduce. Here,
-  // we just dump tensors and their parameter indices into rebuilt_params_ and
-  // rebuilt_param_indices_ based on gradient arriving order, and then at the
-  // end of finalize_backward(), buckets will be rebuilt based on
-  // rebuilt_params_ and rebuilt_param_indices_, and then will be broadcasted
-  // and initialized. Also we only need to dump tensors and parameter indices of
-  // one replica.
-  push_rebuilt_params(index);
+  if (static_graph_ && num_iterations_ == 1) {
+    numGradHooksTriggeredMap_[index] += 1;
+    return;
+  }
 
   // If `find_unused_parameters_` is true there may be model parameters that
   // went unused when computing the model output, they won't be part of the
   // autograd graph, and won't receive gradients. These parameters are
   // discovered in the `prepare_for_backward` function and their indexes stored
   // in the `unused_parameters_` vector.
-  if (!has_marked_unused_parameters_ && find_unused_parameters_) {
+  if (!has_marked_unused_parameters_) {
```

Comment:
Won't removing the `&& find_unused_parameters_` check here change things for regular non static graph training? 

For example if !static_graph and find_unused_parameters=False, this will call mark_variable_ready on each unused index, although I guess it should be a no-op since that list should be empty for this case?

Reply:
yes, for the case '!static_graph and find_unused_parameters=False', unused_parameters_ is empty and thus it is no-op. 

 I removed this check, because in static graph training mode, 'unused_parameters_' is always detected  no matter find_unused_parameters is true or false.